jmxtrans-output/jmxtrans-output-cloudwatch/src/main/java/com/googlecode/jmxtrans/model/output/CloudWatchWriter2.java

```diff
@@ -0,0 +1,141 @@
+import com.google.common.collect.ImmutableList;
+import com.googlecode.jmxtrans.exceptions.LifecycleException;
+import com.googlecode.jmxtrans.model.OutputWriterAdapter;
+import com.googlecode.jmxtrans.model.Query;
+import com.googlecode.jmxtrans.model.Result;
+import com.googlecode.jmxtrans.model.Server;
+import com.googlecode.jmxtrans.model.naming.KeyUtils;
+import com.googlecode.jmxtrans.util.ObjectToDouble;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import java.util.Map;
+
+
+/**
+ * Writes data to <a href="http://aws.amazon.com/cloudwatch/">AWS CloudWatch</a> using the AWS Java SDK
+ *
+ * @author <a href="mailto:sascha.moellering@gmail.com">Sascha Moellering</a>
+ */
+public class CloudWatchWriter2 extends OutputWriterAdapter {
+	private static final Logger log = LoggerFactory.getLogger(CloudWatchWriter2.class);
+
+
+	private final String namespace;
+	private final AmazonCloudWatch cloudWatchClient;
+	private final ObjectToDouble toDoubleConverter = new ObjectToDouble();
+
+	private final ImmutableCollection<Dimension> dimensions;
+	private final ImmutableCollection<String> typeNames;
+
+	public CloudWatchWriter2(
+			@Nonnull String namespace,
+			@Nonnull AmazonCloudWatch cloudWatchClient,
+			@Nonnull ImmutableCollection<Dimension> dimensions,
+			ImmutableCollection<String> typeNames) {
+		this.namespace = namespace;
+		this.cloudWatchClient = cloudWatchClient;
+		this.dimensions = dimensions;
+		this.typeNames = typeNames;
+	}
+
+	@Override
+	public void doWrite(Server server, Query query, Iterable<Result> results) throws Exception {
+		PutMetricDataRequest metricDataRequest = new PutMetricDataRequest();
+		metricDataRequest.setNamespace(namespace);
+		List<MetricDatum> metricDatumList = new ArrayList<>();
+
+		// Iterating through the list of query results
+		for (Result result : results) {
+			try {
+				metricDatumList.add(processResult(result));
+			} catch (IllegalArgumentException iae) {
+				log.error("Could not convert result to double", iae);
+			}
+		}
+
+		metricDataRequest.setMetricData(metricDatumList);
+		cloudWatchClient.putMetricData(metricDataRequest);
+	}
+
+	private MetricDatum processResult(Result result) {
+		// Sometimes the attribute name and the key of the value are the same
+		MetricDatum metricDatum = new MetricDatum();
+		if (result.getValuePath().isEmpty()) {
+			metricDatum.setMetricName(result.getAttributeName());
+		} else {
+			metricDatum.setMetricName(result.getAttributeName() + "_" + KeyUtils.getValuePathString(result));
+		}
+
+		metricDatum.setDimensions(FluentIterable.from(this.dimensions)
+				.append(convertTypeNamesToDimensions(result.getTypeNameMap()))
+				.toList());
+
+		// Converts the Objects to Double-values for CloudWatch
+		metricDatum.setValue(toDoubleConverter.apply(result.getValue()));
+		metricDatum.setTimestamp(new Date());
```

Comment:
Related metricdatum (collected together) may not have the same timestamp

Reply:
@gquintana 
I have tried to keep the functionality same to what it is as of now with `master`. Just added a support for typename in the cloud watch writer. This is from where I referred this code. https://github.com/jmxtrans/jmxtrans/blob/d9ebad93e1d05d298ad159a77a832abaf65b23f3/jmxtrans-output/jmxtrans-output-cloudwatch/src/main/java/com/googlecode/jmxtrans/model/output/CloudWatchWriter.java#L163

However, if you can help me understand what improvement can be made here, I will incorporate the same.